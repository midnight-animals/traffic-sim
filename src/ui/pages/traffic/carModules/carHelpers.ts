export const CAR_HEIGHT = 20;
export const CAR_HEIGHT_FACTOR = 0.3;
export const CAR_GAP = 30;
export const CAR_WIDTH = 50;
export const CAR_ACCELERATION = 3;
export const CHANGE_LANE_SPEED = 3;

/**
 * Cars within range in the front and in the rear = "nearby"
 */
export const NEARBY_VALUE = CAR_WIDTH / 2;

export function getCarHeight(laneRect: DOMRect) {
  return CAR_HEIGHT || laneRect.height * CAR_HEIGHT_FACTOR;
}
