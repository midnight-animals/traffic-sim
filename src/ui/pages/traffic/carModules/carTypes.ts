export interface ICar extends Partial<DOMRect> {
  speed?: number;
  /**
   * Which lane to change to, else `null`
   */
  wantToChangeLane?: number;
}
