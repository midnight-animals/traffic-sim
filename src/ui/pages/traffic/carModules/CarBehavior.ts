import { ICar } from "./carTypes";

/**
 * TODO
 * - Add more behaviors
 *   - change lanes
 *     - if other lane (seems to be) moving faster
 *     - if car in front is moving too slowly
 *     - (if needs to turn)
 */
export class CarBehavior {
  public isAccelerating: boolean;
  public isChangingLanes: boolean;
  public isParking: boolean;

  constructor(private car: ICar) {}

  public setIsAccelerating(is: boolean): void {
    this.isAccelerating = is;
  }
  public setIsChangingLanes(is: boolean): void {
    this.isChangingLanes = is;
  }
  public setIsParking(is: boolean): void {
    this.isParking = is;
  }

  public wantsToAccelerate(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsParking(true);
    }
    return wants;
  }

  public wantsToChangeLane(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsChangingLanes(true);
    }
    return wants;
  }

  public wantsToPark(): boolean {
    const wants = Math.random() >= 0.95;
    if (wants) {
      this.setIsParking(true);
    }
    return wants;
  }
}
