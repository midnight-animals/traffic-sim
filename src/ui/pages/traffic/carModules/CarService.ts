import { getCenterOfCarInLane, getLaneRect } from "../roadModules/roadhelpers";
import { RoadConfig } from "../trafficTypes";
import { CHANGE_LANE_SPEED } from "./carHelpers";
import { ICar } from "./carTypes";

export class CarService {
  private context: CanvasRenderingContext2D;

  constructor(public canvas: HTMLCanvasElement) {
    this.context = canvas.getContext("2d");
  }

  // Function to update the position of a car
  accelerate(car: ICar): void {
    car.x += car.speed;
    if (car.x > this.canvas.width) {
      car.x = -20; // Reset the car's position to the start when it reaches the end of the canvas
    }
  }

  slowDown(car: ICar): void {
    throw new Error("Method not implemented.");
  }

  move(car: ICar, roadConfig: RoadConfig): void {
    this.accelerate(car);
    // car.behavior.wantsToChangeLane() && this.changeLane(car, roadConfig);
    car.behavior.wantsToChangeLane();
    const toLane = this.getLaneToChangeTo(car, roadConfig);
    if (this.canChangeLanes(car, toLane, roadConfig.cars))
      if (car.behavior.isChangingLanes) {
        this.changeLanes(car, toLane);
      }
  }

  /**
   * TODO:
   * - check if `to` lane exists
   */
  canChangeLanes(car: ICar, toLane: number, cars: ICar[]): boolean {
    /* prettier-ignore */ console.log('>>>> _ >>>> ~ cars:', cars);
    return false;
  }

  changeLanes(car: ICar, to: number): void {
    const laneRect = getLaneRect(this.canvas, to);
    const center = getCenterOfCarInLane(to, laneRect);
    const sign = car.lane < to ? 1 : -1;

    if (sign > 0) {
      if (car.y >= center) {
        car.lane = to;
        car.behavior.setIsChangingLanes(false);
        return; // Already in the correct lane
      }
    } else {
      if (car.y <= center) {
        car.lane = to;
        car.behavior.setIsChangingLanes(false);
        return; // Already in the correct lane
      }
    }

    car.y += CHANGE_LANE_SPEED * sign;

    if (car.y > this.canvas.height) {
      car.y = -20; // Reset the car's position to the start when it reaches the end of the canvas
    }
  }

  getLaneToChangeTo(car: ICar, roadConfig: RoadConfig): number {
    const currentLane = car.lane;
    const numOfLanes = roadConfig.lanes.length;
    // Cases at borders
    if (currentLane === numOfLanes) {
      return currentLane - 1;
    } else if (currentLane === 0) {
      return 1;
    }

    return 0;
  }
}
