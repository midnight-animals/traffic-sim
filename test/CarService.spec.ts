import { CarService } from "../src/ui/pages/traffic/carModules/CarService";

function mockWindow() {
  // @ts-expect-error mock window
  global.window = global;
  // @ts-expect-error mock document
  global.document = {
    createElement: jest.fn().mockReturnValue({
      getContext: jest.fn()
    })
  };
}

/**
 * _________________________
 *   [  "1" 11  ]   [  "2" 12  ]
 * ----   ----- -----  -----
 *   [  "3" 21  ]
 * _________________________
 */
const cars_2lanes_2_1 = [
  {
    id: "1",
    x: 0,
    y: 15.5,
    width: 50,
    height: 20,
    speed: 0,
    lane: 0,
    behavior: {} as any
  },
  {
    id: "2",
    x: 80,
    y: 15.5,
    width: 50,
    height: 20,
    speed: 0,
    lane: 0,
    behavior: {} as any
  },
  {
    id: "3",
    x: 0,
    y: 66.5,
    width: 50,
    height: 20,
    speed: 0,
    lane: 1,
    behavior: {} as any
  }
];

describe("CarService", () => {
  let carService: CarService;

  beforeEach(() => {
    // Define window and document for Jest
    mockWindow();
    carService = new CarService(window.document.createElement("canvas"));
  });

  describe("changeLanes", () => {
    test("should change lanes when car is not in the correct lane", () => {
      // carService.changeLanes(car, 2);
    });

    test("should not change lanes when car is already in the correct lane", () => {
      // carService.changeLanes(car, 2);
    });

    test("should reset car position when it reaches the end of the canvas", () => {
      // carService.changeLanes(car, 2);
    });
  });

  describe("canChangeLanes", () => {
    describe("Can change lane", () => {
      test.only("No cars nearby ISSUE-3AFEokbB", () => {
        const car_1_1 = cars_2lanes_2_1[1];
        const result = carService.canChangeLanes(car_1_1, 1, cars_2lanes_2_1);
        const snapshot = {
          car: car_1_1,
          cars: cars_2lanes_2_1,
          result
        };
        expect(snapshot).toMatchSnapshot();
      });
    });

    describe("Cannot change lane", () => {
      test("Car nearby in different lane ISSUE-SlsXBH9D", () => {
        const car_2_1 = cars_2lanes_2_1[2];
        const result = carService.canChangeLanes(car_2_1, 0, cars_2lanes_2_1);
        const snapshot = {
          car: car_2_1,
          cars: cars_2lanes_2_1,
          result
        };
        expect(snapshot).toMatchSnapshot();
      });
    });
  });

  describe("getNearbyCars", () => {
    describe("Nearby", () => {
      test("Same position on different lane ISSUE-Dsc2eTtk", () => {
        const car_1_1 = cars_2lanes_2_1[2];
        const result = carService.getNearbyCars(car_1_1, cars_2lanes_2_1);
        const snapshot = {
          car: car_1_1,
          cars: cars_2lanes_2_1,
          result
        };
        expect(snapshot).toMatchSnapshot();
      });
    });
    describe("Not nearby", () => {
      test("front in different lane ISSUE-uiNjBIP2", () => {
        const car_1_1 = cars_2lanes_2_1[1];
        const result = carService.getNearbyCars(car_1_1, cars_2lanes_2_1);
        const snapshot = {
          car: car_1_1,
          cars: cars_2lanes_2_1,
          result
        };
        expect(snapshot).toMatchSnapshot();
      });
    });
  });
});
