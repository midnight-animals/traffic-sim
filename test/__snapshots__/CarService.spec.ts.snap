// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarService canChangeLanes Can change lane No cars nearby ISSUE-3AFEokbB 1`] = `
{
  "car": {
    "behavior": {},
    "height": 20,
    "id": "2",
    "lane": 0,
    "speed": 0,
    "width": 50,
    "x": 80,
    "y": 15.5,
  },
  "cars": [
    {
      "behavior": {},
      "height": 20,
      "id": "1",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "2",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 80,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "3",
      "lane": 1,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 66.5,
    },
  ],
  "result": true,
}
`;

exports[`CarService canChangeLanes Cannot change lane Car nearby in different lane ISSUE-SlsXBH9D 1`] = `
{
  "car": {
    "behavior": {},
    "height": 20,
    "id": "3",
    "lane": 1,
    "speed": 0,
    "width": 50,
    "x": 0,
    "y": 66.5,
  },
  "cars": [
    {
      "behavior": {},
      "height": 20,
      "id": "1",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "2",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 80,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "3",
      "lane": 1,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 66.5,
    },
  ],
  "result": false,
}
`;

exports[`CarService getNearbyCars Nearby Same position on different lane ISSUE-Dsc2eTtk 1`] = `
{
  "car": {
    "behavior": {},
    "height": 20,
    "id": "3",
    "lane": 1,
    "speed": 0,
    "width": 50,
    "x": 0,
    "y": 66.5,
  },
  "cars": [
    {
      "behavior": {},
      "height": 20,
      "id": "1",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "2",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 80,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "3",
      "lane": 1,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 66.5,
    },
  ],
  "result": [
    {
      "behavior": {},
      "height": 20,
      "id": "1",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 15.5,
    },
  ],
}
`;

exports[`CarService getNearbyCars Not nearby front in different lane ISSUE-uiNjBIP2 1`] = `
{
  "car": {
    "behavior": {},
    "height": 20,
    "id": "2",
    "lane": 0,
    "speed": 0,
    "width": 50,
    "x": 80,
    "y": 15.5,
  },
  "cars": [
    {
      "behavior": {},
      "height": 20,
      "id": "1",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "2",
      "lane": 0,
      "speed": 0,
      "width": 50,
      "x": 80,
      "y": 15.5,
    },
    {
      "behavior": {},
      "height": 20,
      "id": "3",
      "lane": 1,
      "speed": 0,
      "width": 50,
      "x": 0,
      "y": 66.5,
    },
  ],
  "result": [],
}
`;
